### PR TITLE
Correctly zero structures.

### PR DESCRIPTION
**Pull request by paulidale**
Structures that are initialised using `= { 0 }` do not necessarily get fully initialised.  Padding and alignment bytes can be skipped and the C89 standard is ambiguous about fields apart from the first.  Instead, initialise using `memset(3)` to be certain.

Refer to discussion in #8603 for details.

- [x] tests are added or updated

### CODE REVIEW

**Comment by richsalz** [comment]
I think most of this is not necessary.  Only when the C code might look at the structure as an array of bytes, such as memcmp/memcpy/etc or when using as an RNG seed, does this seem necessary. If all access is just via the structure field names, then the code is efficient safe and portable as-is.

**Comment by paulidale** [comment]
@richsalz agreed, I suspect none of these changes are necessary.  However, better to be safe than surprised.  We've been surprised by this already this week.

Still, if the reviewers think I've over done things, I'm happy to drop this PR.

**Comment by levitte** [comment]
I believe it's safe to drop this PR. We would most likely have complaints from the same checkers that complained about the structures used as seed sources otherwise.

Could this somehow be labeled "to be revisited"?

File: include/internal/ktls.h [hunk 1]
```diff
@@ -118,12 +118,13 @@ static ossl_inline int ktls_start(int fd,
 static ossl_inline int ktls_send_ctrl_message(int fd, unsigned char record_type,
                                               const void *data, size_t length)
 {
-    struct msghdr msg = { 0 };
+    struct msghdr msg;
     int cmsg_len = sizeof(record_type);
     struct cmsghdr *cmsg;
     char buf[CMSG_SPACE(cmsg_len)];
```
Comment (bernd-edlinger) [inline include/internal/ktls.h]: the only issue here, is that buf is not guaranteed to be aligned.
I think in another PR, we started to make buf a union with a struct cmsghdr

**Comment by bernd-edlinger** [comment]
I agree, memset should only be necessary when all bytes in a structure are accessed,
like in the RNG.

File: include/internal/ktls.h (hunk 1 above)
Comment (paulidale) [reply include/internal/ktls.h]: Definitely a different PR.  This one is deliberately restricted in scope.

**Comment by paulidale** [comment]
I didn't check if all usages filled in all fields.  I will when I get a chance.

The `msghdr` depends on the structure definition & the kernel.
DISPLAY_COLUMNS could do with a second zero.
The last two need deeper investigation.

**Comment by paulidale** [comment]
I'm happy to strip out the less relevant commits.

The non-init of the struct stat is probably relevant.
I think the DISPLAY_COLUMNS one too.

I'm unsure about the log_counts or PACKET ones -- happy defer these for later consideration.

The array and msghr are fine to lose.

**Comment by t-j-h** [comment]
Actually I tend to err towards safety - in that memset of stack allocated structures are a prudent thing to do. It avoids false positives if the non-defined areas are accessed via anyone who copies the structures or references the structure as a block. It does not hurt to do this in the sense that it will never break something - but it can sometimes be the source of a breakage - and tracking those down certainly consumes more resources than adding a quick memset does.

**Comment by richsalz** [comment]
@t-j-h your safety comment changed my mind and I am in favor of this now.

**Comment by t-j-h** [comment]
@richsalz it is good to have these discussions so the different views are discussed.
Thanks for raising the topic.

**Comment by levitte** [comment]
@t-j-h makes a good point.

**Comment by paulidale** [comment]
Merged, thanks all for the discussion.
We can revert any that aren't suitable if required.

**Pull request closed without merging**